addons/binding/org.openhab.binding.xiaomivacuum/src/main/java/org/openhab/binding/xiaomivacuum/discovery/VacuumDiscovery.java

```diff
@@ -0,0 +1,173 @@
+
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.xiaomivacuum.XiaomiVacuumBindingConstants;
+import org.openhab.binding.xiaomivacuum.internal.Message;
+import org.openhab.binding.xiaomivacuum.internal.Utils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link VacuumDiscovery} is responsible for discovering new Xiaomi Robot Vacuums
+ * and their token
+ *
+ * @author Marcel Verpaalen - Initial contribution
+ *
+ */
+public class VacuumDiscovery extends AbstractDiscoveryService {
+
+    /** The refresh interval for background discovery */
+    private static final long SEARCH_INTERVAL = 600;
+    private static final int TIMEOUT = 2000;
+    private static final int PORT = 54321;
+    private ScheduledFuture<?> roboDiscoveryJob;
+    private Runnable roboDiscoveryRunnable = new Runnable() {
+        @Override
+        public void run() {
+            discover();
+        }
+    };
+    private final static Logger logger = LoggerFactory.getLogger(VacuumDiscovery.class);
+
+    public VacuumDiscovery() throws IllegalArgumentException {
+        super(15);
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        logger.debug("Start Xiaomi Robot Vacuum background discovery");
+        if (roboDiscoveryJob == null || roboDiscoveryJob.isCancelled()) {
+            roboDiscoveryJob = scheduler.scheduleWithFixedDelay(roboDiscoveryRunnable, 0, SEARCH_INTERVAL,
+                    TimeUnit.SECONDS);
+        }
+    }
+
+    @Override
+    protected void stopBackgroundDiscovery() {
+        logger.debug("Stop Xiaomi Robot Vacuum background discovery");
+        if (roboDiscoveryJob != null && !roboDiscoveryJob.isCancelled()) {
+            roboDiscoveryJob.cancel(true);
+            roboDiscoveryJob = null;
+        }
+    }
+
+    private synchronized void discover() {
+        TreeSet<String> broadcastAdresses = getBroadcastAddresses();
+        HashMap<String, byte[]> responses = new HashMap<String, byte[]>();
+        for (String broadcastAdress : broadcastAdresses) {
+            responses.putAll(sendDiscoveryRequest(broadcastAdress));
+        }
+        for (Entry<String, byte[]> i : responses.entrySet()) {
+            logger.debug("Discovery responses from : {}:{}", i.getKey(), Utils.getHex(i.getValue()));
+            Message msg = new Message(i.getValue());
+            String token = Utils.getHexN(msg.getSerialByte());
+            String id = Utils.getHexN(msg.getSerialByte());
+            ThingUID uid = new ThingUID(XiaomiVacuumBindingConstants.THING_TYPE_VACUUM, id);
+            logger.debug("Discovered Xiaomi Robot Vacuum {} at {}", id, i.getKey());
+            if (token.equals("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF")) {
+                logger.debug(
```

Comment:
As this will happen each 5 min in background discovery I was even considering opposite and make it trace.
I did improve the communication around this, if the token is not discovered, it goes offline with a message around the wrong token.